python/ray/train/v2/_internal/execution/worker_group/state.py

```diff
@@ -0,0 +1,137 @@
+import logging
+from dataclasses import dataclass
+from typing import List
+
+import ray
+from ray.actor import ActorHandle
+from ray.train.v2._internal.execution.checkpoint.sync_actor import SynchronizationActor
+from ray.train.v2._internal.execution.worker_group.worker import Worker
+from ray.util.placement_group import PlacementGroup, remove_placement_group
+
+logger = logging.getLogger(__name__)
+
+
+class _WorkerGroupShutdownMixin:
+    """Mixin class for worker group shutdown logic."""
+
+    def _shutdown_workers(self, workers: List[Worker], patience_s: float = 5):
+        # Run the worker shutdown logic on each of the workers. This should
+        # be a non-blocking call to realize forceful shutdown after patience_s.
+        _ = [w.actor.shutdown.remote() for w in workers]
+
+        logger.debug(f"Shutting down {len(workers)} workers.")
+        if patience_s <= 0:
+            for worker in workers:
+                ray.kill(worker.actor)
+        else:
+            done_refs = [w.actor.__ray_terminate__.remote() for w in workers]
+            # Wait for actors to die gracefully.
+            _, not_done = ray.wait(
+                done_refs, num_returns=len(done_refs), timeout=patience_s
+            )
+            if not_done:
+                logger.debug("Graceful termination failed. Falling back to force kill.")
+                # If all actors are not able to die gracefully, then kill them.
+                for worker in workers:
+                    ray.kill(worker.actor)
+
+    def _shutdown_sync_actor(self, sync_actor: SynchronizationActor):
+        ray.kill(sync_actor)
+
+    def _shutdown_placement_group(self, placement_group: PlacementGroup):
+        remove_placement_group(placement_group)
+
+
+@dataclass(frozen=True)
+class WorkerGroupState(_WorkerGroupShutdownMixin):
+    """Ongoing state of an active worker group.
+
+    Attributes:
+        start_time: The time when the worker group was started.
+        workers: The workers in the worker group.
+            These should always be in sorted order by world rank.
+        placement_group: The placement group for the worker group.
+        sync_actor: The synchronization actor for the worker group.
+    """
+
+    start_time: float
+    placement_group: PlacementGroup
+    workers: List[Worker]
+    sync_actor: ActorHandle
+
+    def shutdown(self):
+        self._shutdown_workers(
+            self.workers,
+        )
+        self._shutdown_placement_group(self.placement_group)
+        self._shutdown_sync_actor(self.sync_actor)
+
+
+class WorkerGroupStateBuilder(_WorkerGroupShutdownMixin):
```

Comment:
I think this makes sense to capture all this state that is initialized one by one.

Just a thought -- do you think it makes sense to just have a single class that has a `start` method which attaches the `start_time` and sets a `started` flag to True that signals that the worker group is active?

Reply:
Reason is that the mixin seems a bit unnecessary to me.

Reply:
We can also just not have this logic be a mixin since the utility methods don't require any instance attributes! That's probably the cleanest.